drivers/tee/optee/ocall.c

```diff
@@ -0,0 +1,387 @@
+
+static int verify_request(u32 num_params, struct optee_call_ctx *call_ctx)
+{
+	size_t n;
+	struct optee_msg_arg *arg = call_ctx->rpc_arg;
+
+	switch (arg->cmd) {
+	case OPTEE_MSG_RPC_CMD_SHM_ALLOC:
+	case OPTEE_MSG_RPC_CMD_SHM_FREE:
+		if (!num_params)
+			return -EINVAL;
+
+		/*
+		 * This parameter either carries the requested allocation size
+		 * or a pointer to the SHM to be freed.
+		 */
+		if (!arg->num_params ||
+		    arg->params[0].attr != OPTEE_MSG_ATTR_TYPE_VALUE_INPUT)
+			return -EINVAL;
+		break;
+	case OPTEE_MSG_RPC_CMD_OCALL:
+		/* 'num_params' is checked later */
+
+		/* These parameters carry the OCALL descriptors */
+		if (arg->num_params < 2 ||
+		    arg->params[0].attr != OPTEE_MSG_ATTR_TYPE_VALUE_INOUT ||
+		    arg->params[1].attr != OPTEE_MSG_ATTR_TYPE_VALUE_INPUT ||
+		    arg->params[0].u.value.a > U32_MAX ||  /* OCALL Cmd Id */
+		    arg->params[1].u.value.c != 0)  /* TA UUID (128 bytes) */
+			return -EINVAL;
+
+		/* Skip the loop below */
+		return 0;
+	default:
+		return -EINVAL;
+	}
+
+	for (n = 1; n < arg->num_params; n++)
+		if (arg->params[n].attr != OPTEE_MSG_ATTR_TYPE_NONE)
+			return -EINVAL;
+
+	return 0;
+}
+
+static int verify_reply(u64 func, struct tee_param *params, u32 num_params,
+			struct optee_call_ctx *call_ctx)
+{
+	size_t n;
+
+	switch (func) {
+	case TEE_IOCTL_OCALL_CMD_SHM_ALLOC:
+		if (call_ctx->rpc_arg->cmd != OPTEE_MSG_RPC_CMD_SHM_ALLOC ||
+		    !num_params)
+			return -EINVAL;
+
+		/* This parameter carries the allocated SHM ID */
+		if (params[0].attr != TEE_IOCTL_PARAM_ATTR_TYPE_VALUE_INOUT ||
+		    params[0].u.value.c > INT_MAX)
+			return -EINVAL;
+		break;
+	case TEE_IOCTL_OCALL_CMD_SHM_FREE:
+		if (call_ctx->rpc_arg->cmd != OPTEE_MSG_RPC_CMD_SHM_FREE ||
+		    !num_params)
+			return -EINVAL;
+
+		/* Sanity check, not used while processing the reply */
+		if (params[0].attr != TEE_IOCTL_PARAM_ATTR_TYPE_VALUE_INPUT)
+			return -EINVAL;
+		break;
+	case TEE_IOCTL_OCALL_CMD_INVOKE:
+		if (call_ctx->rpc_arg->cmd != OPTEE_MSG_RPC_CMD_OCALL)
+			return -EINVAL;
+
+		/* Skip the loop below */
+		return 0;
+	default:
+		return -EINVAL;
+	}
+
+	for (n = 1; n < num_params; n++)
+		if (params[n].attr != TEE_IOCTL_PARAM_ATTR_TYPE_NONE)
+			return -EINVAL;
+
+	return 0;
+}
+
+void optee_ocall_process_memrefs(struct optee_msg_param *params, u32 num_params,
+				 bool increment)
+{
+	size_t n;
+
+	for (n = 0; n < num_params; n++) {
+		struct tee_shm *shm;
+		const struct optee_msg_param *mp = params + n;
+		u32 attr = mp->attr & OPTEE_MSG_ATTR_TYPE_MASK;
+
+		switch (attr) {
+		case OPTEE_MSG_ATTR_TYPE_TMEM_INPUT:
+		case OPTEE_MSG_ATTR_TYPE_TMEM_OUTPUT:
+		case OPTEE_MSG_ATTR_TYPE_TMEM_INOUT:
+			shm = (struct tee_shm *)(uintptr_t)mp->u.tmem.shm_ref;
+			break;
+		case OPTEE_MSG_ATTR_TYPE_RMEM_INPUT:
+		case OPTEE_MSG_ATTR_TYPE_RMEM_OUTPUT:
+		case OPTEE_MSG_ATTR_TYPE_RMEM_INOUT:
+			shm = (struct tee_shm *)(uintptr_t)mp->u.rmem.shm_ref;
+			break;
+		default:
+			shm = NULL;
```

Comment:
I'd like to suggest parity in behavior between invoking a function on a TA from a CA and invoking a function from on a CA from a TA, for consistency's sake. If I were programming against this new API, the broken symmetry would stick out.

What are the advantages of mimicking `TEE_InvokeTACommand` regarding this functionality in your view?

It was nevertheless the case that thus far NULL buffer memrefs weren't handled correctly, so in this new iteration I fixed that, under the assumption we'd keep the functionality (and also to see what it would take; most of the heavy lifting is actually in `libteec`).